xbmc/cores/VideoPlayer/DVDCodecs/Overlay/DVDOverlayCodecTX3G.cpp

```diff
@@ -254,31 +253,33 @@ int CDVDOverlayCodecTX3G::Decode(DemuxPacket *pPacket)
     }
   }
 
-  // add a new text element to our container
-  m_pOverlay->AddElement(new CDVDOverlayText::CElementText(strUTF8.c_str()));
+  // We have to remove all \r because it causes the line to display empty box "tofu"
+  StringUtils::Replace(strUTF8, "\r", "");
 
+  m_adapter->AddSubtitle(strUTF8.c_str(), m_pOverlay->iPTSStartTime, m_pOverlay->iPTSStopTime);
+
+  m_output = true;
   return OC_OVERLAY;
 }
 
 void CDVDOverlayCodecTX3G::Reset()
 {
-  if (m_pOverlay)
-    SAFE_RELEASE(m_pOverlay);
+  Dispose();
+  Flush();
 }
 
 void CDVDOverlayCodecTX3G::Flush()
 {
-  if (m_pOverlay)
-    SAFE_RELEASE(m_pOverlay);
+  m_adapter->Flush();
+  m_output = false;
 }
 
 CDVDOverlay* CDVDOverlayCodecTX3G::GetOverlay()
 {
-  if (m_pOverlay)
+  if (m_output)
```

Comment:
Can you explain why you need this here to control the state? Is it to avoid concurrency? Wouldn't a mutex::try_lock be better?

Reply:
i have replicated the current implementation of ass (already present in DVDOverlayCodecSSA) also in the other codecs that now use ass, so i have kept it as
i do not know if change it with a mutex could cause some type of stalls to videoplayersubtitle

Reply:
as discussed need to be investigated and maybe changed in separated PR